primitives/npos-elections/compact/src/codec.rs

```diff
@@ -0,0 +1,205 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Code generation for the ratio assignment type' encode/decode impl.
+
+use crate::field_name_for;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+
+pub(crate) fn codec_impl(
+	ident: syn::Ident,
+	voter_type: syn::Type,
+	target_type: syn::Type,
+	weight_type: syn::Type,
+	count: usize,
+) -> TokenStream2 {
+	let encode = encode_impl(ident.clone(), count);
+	let decode = decode_impl(ident, voter_type, target_type, weight_type, count);
+
+	quote! {
+		#encode
+		#decode
+	}
+}
+
+fn decode_impl(
+	ident: syn::Ident,
+	voter_type: syn::Type,
+	target_type: syn::Type,
+	weight_type: syn::Type,
+	count: usize,
+) -> TokenStream2 {
+	let decode_impl_single = {
+		let name = field_name_for(1);
+		quote! {
+			let #name =
+			<
+				Vec<(_phragmen::codec::Compact<#voter_type>, _phragmen::codec::Compact<#target_type>)>
+				as
+				_phragmen::codec::Decode
+			>::decode(value)?;
+			let #name = #name
+				.into_iter()
+				.map(|(v, t)| (v.0, t.0))
+				// .map(|(v, t) (v.0, t.0)|) // Will crash the compiler in a weird way. Report
+				.collect::<Vec<_>>();
+		}
+	};
+
+	let decode_impl_double = {
+		let name = field_name_for(2);
+		quote! {
+			let #name =
+			<
+				Vec<(
+					_phragmen::codec::Compact<#voter_type>,
+					(_phragmen::codec::Compact<#target_type>, _phragmen::codec::Compact<#weight_type>),
+					_phragmen::codec::Compact<#target_type>,
+				)>
+				as
+				_phragmen::codec::Decode
+			>::decode(value)?;
+			let #name = #name
+				.into_iter()
+				.map(|(v, (t1, w), t2)| (v.0, (t1.0, w.0), t2.0))
+				.collect::<Vec<_>>();
+		}
+	};
+
+	let decode_impl_rest = (3..=count).map(|c| {
+		let name = field_name_for(c);
+
+		let inner_impl = (0..c-1).map(|i|
+			quote! { ( (inner[#i].0).0, (inner[#i].1).0 ), }
+		).collect::<TokenStream2>();
+
+		quote! {
+			let #name =
+			<
+				Vec<(
+					_phragmen::codec::Compact<#voter_type>,
+					[(_phragmen::codec::Compact<#target_type>, _phragmen::codec::Compact<#weight_type>); #c],
```

Comment:
isn't it `#c-1` ? because vote3 is `votes3: Vec<(u32, [(u8, TestAccuracy); 2usize], u8)>,`

```suggestion
					[(_phragmen::codec::Compact<#target_type>, _phragmen::codec::Compact<#weight_type>); #c-1],
```

Reply:
damn 🤦  right, thanks.

Reply:
Luckily it was harmless but could have been a serious bug.